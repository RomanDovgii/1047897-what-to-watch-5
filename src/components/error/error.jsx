import React from "react";
import {connect} from "react-redux";
import {flushError} from "../../store/actions/action";
import {errorType} from "../types/types";

const errorStyle = {
  ERROR: {
    paddingTop: `10px`,
    paddingRight: `10px`,
    paddingBottom: `20px`,
    paddingLeft: `10px`,
    width: `300px`,
    minHeight: `50px`,
    position: `fixed`,
    bottom: `0`,
    right: `30px`,
    backgroundImage: `linear-gradient(-180deg,#180202 0,#000 100%)`,
    boxShadow: `0px 0px 20px 2px #330233`,
    zIndex: `1000`,
    opacity: `0.9`
  },
  ERROR_HEADING: {
    margin: `0`,
    paddingRight: `20px`,
    fontSize: `30px`,
    color: `#ffffff`
  },
  ERROR_TEXT: {
    margin: `0`,
    paddingTop: `20px`,
    color: `#ffffff`
  },
  ERROR_CLOSE: {
    position: `absolute`,
    top: `10px`,
    right: `5px`,
    width: `30px`,
    height: `30px`,
    background: `none`,
    border: `none`,
    display: `flex`,
    flexDirection: `column`,
    alignItems: `center`,
    justifyContent: `center`
  },
  ERROR_CLOSE_BEFORE: {
    content: `\`\``,
    width: `4px`,
    height: `20px`,
    backgroundColor: `#ffffff`,
    position: `absolute`,
    borderRadius: `20px`,
    transform: `rotate(40deg)`
  },
  ERROR_CLOSE_AFTER: {
    content: `\`\``,
    width: `4px`,
    height: `20px`,
    backgroundColor: `#ffffff`,
    position: `absolute`,
    borderRadius: `20px`,
    transform: `rotate(-40deg)`
  },
  ERROR_CLOSE_TEXT: {
    position: `absolute !important`,
    clip: `rect(1px, 1px, 1px, 1px)`,
    padding: `0 !important`,
    border: `0 !important`,
    height: `1px !important`,
    width: `1px !important`,
    overflow: `hidden`
  },
};

const Error = (props) => {
  const {error, isError, onCloseButtonClick} = props;
  const {heading = `Error Unknown`, text = `We don't know what happened. If it doesn't bark, doesn't spread seeds or communism, and doesn't try to kill you, then just ignore it. It isn't important.`} = error;

  return (
    <React.Fragment>
      {isError &&
        <div style={errorStyle.ERROR}>
          <h6 style={errorStyle.ERROR_HEADING}>{heading}</h6>
          <p style={errorStyle.ERROR_TEXT}>
            {text}
          </p>
          <button className="error__close" style={errorStyle.ERROR_CLOSE} onClick={onCloseButtonClick}>
            <div style={errorStyle.ERROR_CLOSE_BEFORE}/>
            <div style={errorStyle.ERROR_CLOSE_AFTER}/>
            <p style={errorStyle.ERROR_CLOSE_TEXT}>Close popup</p>
          </button>
        </div>}
    </React.Fragment>
  );
};

Error.propTypes = errorType;

const mapStateToProps = ({STATE}) => ({
  isError: STATE.isError,
  error: STATE.error
});

const mapDispatchToProps = (dispatch) => ({
  onCloseButtonClick() {
    dispatch(flushError());
  }
});

export {Error};
export default connect(mapStateToProps, mapDispatchToProps)(Error);
