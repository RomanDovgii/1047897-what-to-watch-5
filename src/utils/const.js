export const MOVIES_COUNT = 9;
export const MAXIMUM_RAITING = 10;
export const MINIMUM_STARRING = 5;
export const MAXIMUM_STARRING = 10;
export const MAXIMUM_RAITING_COUNT = 1000;
export const MINIMUM_DESCRIPTION = 1;
export const MAXIMUM_DESCRIPTION = 10;

export const GENRES = [`All genres`, `Comedies`, `Crime`, `Documentary`, `Dramas`, `Horror`, `Kids & Family`, `Romance`, `Sci-Fi`, `Thrillers`];
export const MOVIES = [`Aviator`, `Bohemian Rhapsody`, `Dardjeeling Limited`, `Fantastic Beasts The Crimes of Grindelwald`, `Johnny English`, `Macbeth`, `Midnight Special`, `Mindhunter`, `Moonrise Kingdom`, `No Country For Old Men`, `Orlando`, `Pulp Fiction`, `Revenant`, `Seven Years In Tibet`, `Shutter Island`, `Snatch`, `War of the Worlds`, `We need to talk about Kevin`, `What We Do in the Shadows`];
export const VIDEO_URLS = [`https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`, `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`];
export const DIRECTORS = [`Steven Spielberg`, `Martin Scorsese`, `Alfred Hitchcock`, `Quentin Tarantino`, `Stanley Kubrick`, `Christopher Nolan`, `Tim Burton`, `Spike lee`];
export const ACTORS = [`Tom Hanks`, `Will Smith`, `Leonardo DiCaprio`, `Robert De Niro`, `Tom Cruise`, `Samuel L. Jackson`, `Robert Downey, Jr.`, `Scarlett Johansson`, `Jennifer Lawrence`, `Nickole Kidman`, `Cate Blanchett`, `Meryl Streep`, `Charlize Theron`, `Amy Adams`, `Angelina Jolie`];

export const FISH_TEXT = `Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc sodales convallis elit et lacinia. Quisque quis cursus lectus. Praesent elementum semper felis ut pretium. Ut eu mi mollis, luctus purus ac, aliquet ipsum. Maecenas libero nunc, faucibus a porttitor eget, laoreet hendrerit diam. Aliquam eget purus sit amet augue feugiat efficitur in et augue. Curabitur nec nulla a mauris maximus semper vel in dui. Vivamus mollis urna ex, id facilisis velit facilisis tincidunt. Maecenas vitae congue massa. Maecenas vel tortor justo.`;

export const GRAND_BUDAPEST_HOTEL = {
  name: `The Grand Budapest Hotel`,
  genre: `Drama`,
  release: 2014,
  description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege. Gustave prides himself on providing first-class service to the hotel's guests, including satisfying the sexual needs of the many elderly women who stay there. When one of Gustave's lovers dies mysteriously, Gustave finds himself the recipient of a priceless painting and the chief suspect in her murder.`,
  director: `Wes Andreson`,
  starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`, `Tom Cruise`],
  rating: 8.9,
  ratingsCount: 240,
  videoUrl: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  isMyList: false
};

export const CatalogHeadingVariant = {
  CATALOG: `Catalog`, // скрыт
  MOVIE_PAGE: `More like this` // не скрыт
};
