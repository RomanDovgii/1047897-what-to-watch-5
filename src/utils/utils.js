import React from "react";
import {GENRES, MOVIES, VIDEO_URLS, DIRECTORS, ACTORS, MOVIES_COUNT, MAXIMUM_RATING, MAXIMUM_RATING_COUNT, FISH_TEXT, ActorsCount, Description, Paragraph, Duration, Id, CommentCount, MAXIMUM_SIMIALAR_MOVIES, MAXIMUM_DISPLAYED_ACTORS, RatingSystem, ALL_GENRE} from "./const";

const getRandomElementFromArray = (array) => {
  const max = array.length;
  const randomNumber = Math.floor(Math.random() * max);

  return array[randomNumber];
};

const getRandomIntegerNumber = (firstNumber, secondNumber = 0) => {
  const max = Math.max(firstNumber, secondNumber);
  const min = Math.min(firstNumber, secondNumber);

  return Math.floor(Math.random() * (max - min) + min);
};

const getRandomBoolean = () => {
  return (Math.random() >= 0.5);
};

const generateRating = () => {
  return Math.round((Math.random() * MAXIMUM_RATING) * 1e1) / 1e1;
};

const generateActors = () => {
  const actors = [];

  const starringCount = getRandomIntegerNumber(ActorsCount.MINIMUM, ActorsCount.MAXIMUM);

  for (let i = 0; i < starringCount; i++) {
    actors.push(getRandomElementFromArray(ACTORS));
  }

  return actors;
};

const generateParagraphText = () => {
  const randomSentences = [];
  const sentences = FISH_TEXT.split(`. `);
  const descriptionLength = getRandomIntegerNumber(Description.MINIMUM, Description.MAXIMUM);

  for (let i = 0; i < descriptionLength; i++) {
    randomSentences.push(getRandomElementFromArray(sentences));
  }

  const text = randomSentences.join(`. `);

  return text;
};

const generateDescriptionParagraphs = () => {
  const paragraphsCount = getRandomIntegerNumber(Paragraph.MINIMUM, Paragraph.MAXIMUM);
  let paragraphs = [];

  for (let i = 0; i < paragraphsCount; i++) {
    paragraphs.push(generateParagraphText());
  }

  return paragraphs;
};

const generateRelease = () => {
  const date = new Date();

  return date.getFullYear();
};

const generateMovie = () => {
  return {
    name: getRandomElementFromArray(MOVIES),
    genre: getRandomElementFromArray(GENRES),
    release: generateRelease(),
    descriptionParagraphs: generateDescriptionParagraphs(),
    director: getRandomElementFromArray(DIRECTORS),
    actors: generateActors(),
    rating: generateRating(),
    ratingsCount: getRandomIntegerNumber(MAXIMUM_RATING_COUNT),
    videoUrl: getRandomElementFromArray(VIDEO_URLS),
    duration: getRandomIntegerNumber(Duration.MINIMUM, Duration.MAXIMUM),
    id: getRandomIntegerNumber(Id.MINIMUM, Id.MAXIMUM),
    isMyList: getRandomBoolean()
  };
};

export const generateComment = () => {
  return {
    text: generateParagraphText(),
    rating: generateRating(),
    name: getRandomElementFromArray(ACTORS),
    date: new Date()
  };
};

const generateComments = (count) => {
  return new Array(count).fill().map(generateComment);
};

export const generateCommentsById = (movieIds) => {
  const comments = movieIds.reduce((accumulator, id) => {
    const number = getRandomIntegerNumber(CommentCount.MINIMUM, CommentCount.MAXIMUM);
    accumulator[id] = generateComments(number);
    return accumulator;
  }, {});

  return comments;
};

export const filterMoviesForMyList = (movies) => {
  return movies.filter((movie) => movie.isMyList);
};

export const generateMovies = () => {
  return new Array(MOVIES_COUNT).fill().map(generateMovie);
};

export const generateVideoType = (url) => {
  const urlPieces = url.split(`.`);
  const lastIndex = urlPieces.length - 1;

  return `video/${urlPieces[lastIndex]}`;
};

export const filterMoviesByGenre = (movies, genre) => {
  const filteredMovies = movies.slice().filter((movie) => movie.genre === genre);

  let finalArray = filteredMovies;

  if (filteredMovies.length < MAXIMUM_SIMIALAR_MOVIES) {
    const difference = MAXIMUM_SIMIALAR_MOVIES - filterMoviesByGenre.length;
    const additionalMovies = movies.slice().filter((movie) => movie.genre !== genre).slice(difference);
    finalArray = filteredMovies.concat(additionalMovies);
  }

  return finalArray;
};

export const generateDurationString = (duration) => {
  return `${Math.floor(duration / 60)}h ${duration % 60}m`;
};

export const generateActorsText = (actors) => {
  const actorsMaxIndex = actors.length - 1;

  return (
    <span className="movie-card__details-value">
      {actors.map((actor, i) => (
        <React.Fragment key={i}>
          {actor}{i < actorsMaxIndex ? `, ` : ``} {i < actorsMaxIndex ? <br/> : ``}
        </React.Fragment>
      ))}
    </span>);
};

export const generateStarringString = (actors) => {
  let text = ``;
  let end = ``;
  let actorsLocal = actors.slice();

  if (actorsLocal.length > MAXIMUM_DISPLAYED_ACTORS) {
    actorsLocal = actors.slice(0, MAXIMUM_DISPLAYED_ACTORS);

    end = ` and other`;
  }

  text = actorsLocal.join(`, `);

  text = text.slice(0, -2);

  text += end;

  return text;
};

export const generateRatingText = (rating) => {
  switch (true) {
    case rating <= RatingSystem.BAD:
      return `Bad`;
    case rating <= RatingSystem.NORMAL:
      return `Normal`;
    case rating <= RatingSystem.GOOD:
      return `Good`;
    case rating < RatingSystem.AWESOME:
      return `Very good`;
    case rating === RatingSystem.AWESOME:
      return `Awesome`;
    default:
      return `Rating is incorrect`;
  }
};

export const extend = (a, b) => {
  return Object.assign({}, a, b);
};

export const createGenresList = (movies) => {
  const genresNotSorted = [];

  movies.map((movie) => {
    genresNotSorted.push(movie.genre);
  });

  const genres = [];

  genresNotSorted.map((genre) => {
    if (!genres.includes(genre)) {
      genres.push(genre);
    }
  });

  return [ALL_GENRE, ...genres];
};
