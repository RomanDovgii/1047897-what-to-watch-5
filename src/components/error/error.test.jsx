import React from "react";
import renderer from "react-test-renderer";
import {Error} from "./error";
import {noop} from "../../utils/utils";

const testError = {
  code: `Unknown`,
  text: `He, he, he, try to catch me`
};

it(
    `Should render correct error`,
    () => {
      const tree = renderer
      .create(
          <Error
            error={testError}
            isError={true}
            onCloseButtonClick={noop}
          />
      )
      .toJSON();

      expect(tree).toMatchSnapshot();
    }
);

it(
    `Should render nothing`,
    () => {
      const tree = renderer
  .create(
      <Error
        error={{}}
        isError={false}
        onCloseButtonClick={noop}
      />
  )
  .toJSON();

      expect(tree).toMatchSnapshot();
    }
);

it(
    `Should render nothing even if error exist`,
    () => {
      const tree = renderer
    .create(
        <Error
          error={testError}
          isError={false}
          onCloseButtonClick={noop}
        />
    )
    .toJSON();

      expect(tree).toMatchSnapshot();
    }
);

it(
    `In case of back end developer being not smart enough to provide information with error status (error code 1D10T). It will render an error with basic text`,
    () => {
      const tree = renderer
    .create(
        <Error
          error={{}}
          isError={true}
          onCloseButtonClick={noop}
        />
    )
    .toJSON();

      expect(tree).toMatchSnapshot();
    }
);
