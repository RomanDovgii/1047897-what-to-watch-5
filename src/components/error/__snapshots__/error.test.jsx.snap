// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`In case of back end developer being not smart enough to provide information with error status (error code 1D10T). It will render an error with basic text 1`] = `
<div
  style={
    Object {
      "backgroundImage": "linear-gradient(-180deg,#180202 0,#000 100%)",
      "bottom": "0",
      "boxShadow": "0px 0px 20px 2px #330233",
      "minHeight": "50px",
      "opacity": "0.9",
      "paddingBottom": "20px",
      "paddingLeft": "10px",
      "paddingRight": "10px",
      "paddingTop": "10px",
      "position": "fixed",
      "right": "30px",
      "width": "300px",
      "zIndex": "1000",
    }
  }
>
  <h6
    style={
      Object {
        "color": "#ffffff",
        "fontSize": "30px",
        "margin": "0",
        "paddingRight": "20px",
      }
    }
  >
    Error Unknown
  </h6>
  <p
    style={
      Object {
        "color": "#ffffff",
        "margin": "0",
        "paddingTop": "20px",
      }
    }
  >
    We don't know what happened. If it doesn't bark, doesn't spread seeds or communism, and doesn't try to kill you, then just ignore it. It isn't important.
  </p>
  <button
    className="error__close"
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "background": "none",
        "border": "none",
        "display": "flex",
        "flexDirection": "column",
        "height": "30px",
        "justifyContent": "center",
        "position": "absolute",
        "right": "5px",
        "top": "10px",
        "width": "30px",
      }
    }
  >
    <div
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderRadius": "20px",
          "content": "\`\`",
          "height": "20px",
          "position": "absolute",
          "transform": "rotate(40deg)",
          "width": "4px",
        }
      }
    />
    <div
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderRadius": "20px",
          "content": "\`\`",
          "height": "20px",
          "position": "absolute",
          "transform": "rotate(-40deg)",
          "width": "4px",
        }
      }
    />
    <p
      style={
        Object {
          "border": "0 !important",
          "clip": "rect(1px, 1px, 1px, 1px)",
          "height": "1px !important",
          "overflow": "hidden",
          "padding": "0 !important",
          "position": "absolute !important",
          "width": "1px !important",
        }
      }
    >
      Close popup
    </p>
  </button>
</div>
`;

exports[`Should render correct error 1`] = `
<div
  style={
    Object {
      "backgroundImage": "linear-gradient(-180deg,#180202 0,#000 100%)",
      "bottom": "0",
      "boxShadow": "0px 0px 20px 2px #330233",
      "minHeight": "50px",
      "opacity": "0.9",
      "paddingBottom": "20px",
      "paddingLeft": "10px",
      "paddingRight": "10px",
      "paddingTop": "10px",
      "position": "fixed",
      "right": "30px",
      "width": "300px",
      "zIndex": "1000",
    }
  }
>
  <h6
    style={
      Object {
        "color": "#ffffff",
        "fontSize": "30px",
        "margin": "0",
        "paddingRight": "20px",
      }
    }
  >
    Error Unknown
  </h6>
  <p
    style={
      Object {
        "color": "#ffffff",
        "margin": "0",
        "paddingTop": "20px",
      }
    }
  >
    He, he, he, try to catch me
  </p>
  <button
    className="error__close"
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "background": "none",
        "border": "none",
        "display": "flex",
        "flexDirection": "column",
        "height": "30px",
        "justifyContent": "center",
        "position": "absolute",
        "right": "5px",
        "top": "10px",
        "width": "30px",
      }
    }
  >
    <div
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderRadius": "20px",
          "content": "\`\`",
          "height": "20px",
          "position": "absolute",
          "transform": "rotate(40deg)",
          "width": "4px",
        }
      }
    />
    <div
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderRadius": "20px",
          "content": "\`\`",
          "height": "20px",
          "position": "absolute",
          "transform": "rotate(-40deg)",
          "width": "4px",
        }
      }
    />
    <p
      style={
        Object {
          "border": "0 !important",
          "clip": "rect(1px, 1px, 1px, 1px)",
          "height": "1px !important",
          "overflow": "hidden",
          "padding": "0 !important",
          "position": "absolute !important",
          "width": "1px !important",
        }
      }
    >
      Close popup
    </p>
  </button>
</div>
`;

exports[`Should render nothing 1`] = `null`;

exports[`Should render nothing even if error exist 1`] = `null`;
